### PR TITLE
Store windows in OpenRCT2 memory and increase maximum number of windows

### PR DESCRIPTION
**Pull request by IntelOrca**
Allocate a bigger window list and viewport list in OpenRCT2 static memory and move over all references from RCT2.

This also removes the 'active' viewport pointer list and instead changes everything to just enumerate the whole viewport list

This needs some testing, we must be certain no more RCT2 code that is called enumerates the windows or viewports - unless through one of our hook functions.

image

Having many viewports open is somewhat a performance issue, but that's up to user I suppose.

### CODE REVIEW

**Comment by janisozaur** [comment]
That screenshot, though… and still 26fps? wow!

File: src/world/map.c [hunk 1]
```diff
@@ -4779,12 +4780,11 @@ void map_invalidate_tile_under_zoom(int x, int y, int z0, int z1, int maxZoom)
 	x2 = x + 32;
 	y2 = y + 32 - z0;
 
-	vp = RCT2_ADDRESS(RCT2_ADDRESS_ACTIVE_VIEWPORT_PTR_ARRAY, rct_viewport*);
-	while (!gOpenRCT2Headless && *vp != 0) {
-		if (maxZoom == -1 || (*vp)->zoom <= maxZoom) {
-			viewport_invalidate(*vp, x1, y1, x2, y2);
+	for (int i = 0; i < MAX_VIEWPORT_COUNT; i++) {
+		rct_viewport *viewport = &g_viewport_list[i];
+		if (viewport->width != 0 && maxZoom == -1 || viewport->zoom <= maxZoom) {
```
Comment (janisozaur) [inline src/world/map.c]:
```
/home/janisozaur/workspace/OpenRCT2/src/world/map.c:4785:28: warning: '&&' within '||' [-Wlogical-op-parentheses]
                if (viewport->width != 0 && maxZoom == -1 || viewport->zoom <= maxZoom) {
                    ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~ ~~
/home/janisozaur/workspace/OpenRCT2/src/world/map.c:4785:28: note: place parentheses around the '&&' expression to silence this warning
                if (viewport->width != 0 && maxZoom == -1 || viewport->zoom <= maxZoom) {
                                         ^
                    (                                    )
```

**Comment by janisozaur** [comment]
With that many windows it gets harder to notice frame blinking when you try to open already opened window. Perhaps that should be addressed as well.

**Comment by IntelOrca** [comment]
How?

**Comment by janisozaur** [comment]
flashing more than once, different colour or perhaps flashing more than just 1px frame

**Comment by ledust** [comment]
Thanks you so much this is crazy that you added this exactly as I was talking about it! :D so happy I can monitor all my users now easily!

**Pull request merged**